gk/main.c

```diff
@@ -417,21 +425,141 @@ setup_gk_instance(unsigned int lcore_id, struct gk_config *gk_conf)
 		goto flow_hash;
 	}
 
+	ret = snprintf(ring_name, sizeof(ring_name), "gk_ring_%d", block_idx);
+	RTE_ASSERT(ret < sizeof(ring_name));
+
+	instance->cmd_ring = (struct rte_ring *)rte_ring_create(ring_name,
+		gk_conf->flow_ht_size / 2, socket_id, RING_F_SC_DEQ);
+    	if (instance->cmd_ring == NULL) {
+		RTE_LOG(ERR, RING,
+			"The GK block can't create ring at lcore %u!\n",
+			lcore_id);
+
+		ret = -1;
+        	goto flow_entry;
+    	}
+
+	ret = snprintf(pool_name, sizeof(pool_name), "gk_pool_%d", block_idx);
+	RTE_ASSERT(ret < sizeof(pool_name));
+
+    	instance->cmd_pool = (struct rte_mempool *)rte_mempool_create(pool_name, 
+                gk_conf->flow_ht_size / 2, sizeof(struct gk_cmd_entry),
+                GK_MEM_CACHE_SIZE, 0, NULL, NULL, NULL, NULL, socket_id, 0);
+    	if (instance->cmd_pool == NULL) {
+		RTE_LOG(ERR, MEMPOOL,
+			"The GK block can't create mempool at lcore %u!\n",
+			lcore_id);
+
+		ret = -1;
+        	goto cmd_ring;
+    	}
+
 	ret = 0;
 	goto out;
 
+ 
+cmd_ring:
+    	rte_ring_free(instance->cmd_ring);
+    	instance->cmd_ring = NULL;
+flow_entry:
+    	rte_free(instance->ip_flow_entry_table);
+    	instance->ip_flow_entry_table = NULL;
 flow_hash:
 	rte_hash_free(instance->ip_flow_hash_table);
 	instance->ip_flow_hash_table = NULL;
 out:
 	return ret;
 }
 
+static void
+add_ggu_policy(struct ggu_policy *policy, struct gk_instance *instance)
+{
+	int ret;
+	struct flow_entry *fe;
+
+	ret = rte_hash_lookup_with_hash(instance->ip_flow_hash_table,
+		&policy->flow, rss_ip_flow_hf(&policy->flow, 0, 0));
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"The GK block failed to find the flow key in hash table, so it cannot fill up the flow entry according to the policy decision!\n");
+		return;
+	}
+
+	fe = &instance->ip_flow_entry_table[ret];
+
+	switch(policy->params.state) {
+	case GK_REQUEST:
```

Comment:
I think the grantor should only return the policy decision with either ``granted`` or ``rejected``. Only flows that are not determined should be in ``request`` state. So, we need to remove the ``GK_REQUEST`` statement, right?